File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/di/AppComponent.java
package com.example.android.architecture.blueprints.todoapp.di;

import com.example.android.architecture.blueprints.todoapp.ToDoApplication;
import com.example.android.architecture.blueprints.todoapp.data.source.TasksRepository;
import com.example.android.architecture.blueprints.todoapp.data.source.TasksRepositoryModule;

import javax.inject.Singleton;

import dagger.Component;
import dagger.android.AndroidInjector;
import dagger.android.support.AndroidSupportInjectionModule;

/**
 * This is a Dagger component. Refer to {@link ToDoApplication} for the list of Dagger components
 * used in this application.
 * <p>
 * Even though Dagger allows annotating a {@link Component} as a singleton, the code
 * itself must ensure only one instance of the class is created. This is done in {@link
 * ToDoApplication}.
 * //{@link AndroidSupportInjectionModule}
 * // is the module from Dagger.Android that helps with the generation
 * // and location of subcomponents.
 */
@Singleton
@Component(modules = {TasksRepositoryModule.class,
        ApplicationModule.class,
        ActivityBindingModule.class,
        AndroidSupportInjectionModule.class})
public interface AppComponent extends AndroidInjector<ToDoApplication> {

    TasksRepository getTasksRepository();

    // Application will just be provided into our app graph now.
    @Component.Builder
    abstract class Builder extends AndroidInjector.Builder<ToDoApplication> {
    }
}

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/di/AppComponent.kt
package com.example.android.architecture.blueprints.todoapp.di

import android.content.Context
import com.example.android.architecture.blueprints.todoapp.addedittask.di.AddEditTaskComponent
import com.example.android.architecture.blueprints.todoapp.data.source.TasksRepository
import com.example.android.architecture.blueprints.todoapp.statistics.di.StatisticsComponent
import com.example.android.architecture.blueprints.todoapp.taskdetail.di.TaskDetailComponent
import com.example.android.architecture.blueprints.todoapp.tasks.di.TasksComponent
import dagger.BindsInstance
import dagger.Component
import dagger.Module
import javax.inject.Singleton

/**
 * Main component for the application.
 *
 * See the `TestApplicationComponent` used in UI tests.
 */
@Singleton
@Component(
    modules = [
        AppModule::class,
        AppModuleBinds::class,
        ViewModelBuilderModule::class,
        SubcomponentsModule::class
    ]
)
interface AppComponent {

    @Component.Factory
    interface Factory {
        fun create(@BindsInstance applicationContext: Context): AppComponent
    }

    fun addEditTaskComponent(): AddEditTaskComponent.Factory
    fun statisticsComponent(): StatisticsComponent.Factory
    fun taskDetailComponent(): TaskDetailComponent.Factory
    fun tasksComponent(): TasksComponent.Factory

    val tasksRepository: TasksRepository
}

@Module(
    subcomponents = [
        TasksComponent::class,
        AddEditTaskComponent::class,
        StatisticsComponent::class,
        TaskDetailComponent::class
    ]
)
object SubcomponentsModule